### PR TITLE
Add shr_string_listGetAllNames and associated unit tests

### PR DESCRIPTION
**Pull request by billsacks**
I plan to leverage this new subroutine in the water tracer code: From the config file, we'll read in colon-delimited lists of water tracer names, water tracer species, and water tracer initial ratios; we then will need to turn each of those into arrays of values; this new subroutine accomplishes that step.

I also added some unit tests of shr_string_listIsValid. I was originally planning to modify that routine to add some flexibility in terms of what's considered "valid", but that started to feel messy, so I'm instead planning to keep that as is. But before deciding to leave it as is, I had already written some unit tests to cover my planned rework; I figured I might as well leave them in place to cover this code for the future. If anyone is interested (probably mainly @nusbaume ): The two changes I had been thinking of making to this were (1) optionally allowing 0-length lists (in case we have 0 water tracers), and (2) optionally allowing 0-length elements (in case the species name is empty, indicating bulk water). For (1), I decided to handle this by checking this in the caller and doing special handling of 0-length lists. For (2), I decided to use "-" as the species name for bulk water rather than an empty string so that I could use the shr_string_list functionality unchanged.

### CODE REVIEW

File: src/shr_string_mod.F90 [hunk 1]
```diff
@@ -1011,6 +1012,106 @@ subroutine shr_string_listGetName(list,k,name,rc)
...
+  !     2025-Nov-10 - W. Sacks
+  !
+  ! !INTERFACE: ------------------------------------------------------------------
+
+  subroutine shr_string_listGetAllNames(list,names,rc)
+
+    ! !INPUT/OUTPUT PARAMETERS:
+
+    character(*)                 ,intent(in)  :: list     ! list/string
+    character(*), allocatable    ,intent(out) :: names(:) ! k-th name in list
```
Comment (nusbaume) [inline src/shr_string_mod.F90]: I assume the  comment should be changed here?

```suggestion
    character(*), allocatable    ,intent(out) :: names(:) ! array of all names in list
```

File: src/shr_string_mod.F90 [hunk 2]
```diff
@@ -1011,6 +1012,106 @@ subroutine shr_string_listGetName(list,k,name,rc)
...
+
+    if (debug>1 .and. t01<1) call shr_timer_get(t01,subName)
+    if (debug>1) call shr_timer_start(t01)
+
+    rCode = 0
+
+    !--- check that this is a valid list ---
+    if (.not. shr_string_listIsValid(list,rCode) ) then
+       write(s_logunit,F00) "ERROR: invalid list = ",trim(list)
+       call shr_string_abort(subName//" ERROR: invalid list = "//trim(list))
```
Comment (nusbaume) [inline src/shr_string_mod.F90]: For error messages should we "wrap" the list with some markers to make it clear what exactly is in the list?  For example:

```suggestion
       write(s_logunit,F00) "ERROR: invalid list = <",trim(list),">"
       call shr_string_abort(subName//" ERROR: invalid list = <"//trim(list)//">")
```

I just know that in CAM it can make it easier to interpret error messages, especially when special characters are potentially involved.  Also note that I only chose `<` and `>` because that is what is being used below as well, but am happy if you would prefer different symbols instead.

File: src/shr_string_mod.F90 [hunk 3]
```diff
@@ -1011,6 +1012,106 @@ subroutine shr_string_listGetName(list,k,name,rc)
...
+    end if
+
+    num_fields = shr_string_listGetNum(list)
+
+    allocate(names(num_fields))
```
Comment (nusbaume) [inline src/shr_string_mod.F90:1071]: Not sure what the general standards are for share code, but should we do a `stat` check for the allocate statement here?

File: test/unit/shr_string_test/test_shr_string.pf [hunk 4]
```diff
@@ -12,6 +12,193 @@ module test_shr_string
...
+  subroutine test_shr_string_listGetAllNames_tooLong()
+    ! Test shr_string_listGetAllNames with a too-long element
+    character(len=32) :: list
+    character(len=4), allocatable :: names(:)
+    integer :: rc
+
+    list = "abc:defgh:ijk"
+    call shr_string_listGetAllNames(list, names, rc)
+    @assertExceptionRaised("ABORTED: (shr_string_listGetAllNames)  ERROR: an element of list <abc:defgh:ijk> exceeds the max char length of the output variable")
+  end subroutine test_shr_string_listGetAllNames_tooLong
```
Comment (nusbaume) [inline test/unit/shr_string_test/test_shr_string.pf:200]: Should we add one last test for `shr_string_listGetAllNames` that ensures that a `.false.` return from `shr_string_listIsValid` raises the correct error/message?  Of course if this causes a problem given the logging step then I am happy to skip it for now.

**Review by nusbaume — CHANGES_REQUESTED** [review]
Thanks for the new functionality @billsacks!  I just had some (hopefully minor) requests/questions.

File: src/shr_string_mod.F90 (hunk 1 above)
Comment (billsacks) [reply src/shr_string_mod.F90]: Good catch, thanks! (I swear I didn't start by copying & pasting this code... <looks around innocently>).

File: src/shr_string_mod.F90 (hunk 2 above)
Comment (billsacks) [reply src/shr_string_mod.F90]: Yes, good idea about enclosing the list in "<" and ">".

File: src/shr_string_mod.F90 (hunk 3 above)
Comment (billsacks) [reply src/shr_string_mod.F90:1071]: Good question. Most allocate statements in share do *not* check stat. My personal feeling is that that's okay, and actually preferable in many cases: My understanding/experience is that you'll still get an error message if the allocation runs into problems... yes, you can use stat to give a more elegant / graceful error message, but in cases like this where it seems extremely unlikely to get an error (since the allocation is small), I'm comfortable with just relying on the error message and avoiding the extra error-handling code.

I'm open to revisiting that if you'd like.

File: test/unit/shr_string_test/test_shr_string.pf (hunk 4 above)
Comment (billsacks) [reply test/unit/shr_string_test/test_shr_string.pf:200]: That's a good idea, but I can't get this to work due to limitations in what can and can't be done with expected error tests in pFUnit, or Fortran more generally, due to the lack of exceptions in Fortran.

Here's the test I tried to write:
```Fortran
  @Test
  subroutine test_shr_string_listGetAllNames_invalid()
    ! Test shr_string_listGetAllNames with an invalid list (should abort)
   character(len=32) :: list
   character(len=4), allocatable :: names(:)
   integer :: rc

   ! Note the presence of a 0-length item here; this is invalid
   list = "abc::def"
   call shr_string_listGetAllNames(list, names, rc)
   @assertExceptionRaised("ABORTED: (shr_string_listGetAllNames)  ERROR: invalid list = <abc::def>")
  end subroutine test_shr_string_listGetAllNames_invalid
```

The problem is that, in unit testing, we replace the actual shr_abort with a version that raises a pFUnit "exception" but doesn't actually abort (this is needed because an actual abort would skip the rest of the tests in this test file). So then the code happily continues on within the given routine rather than aborting at the error location like it would in production.

In this case, this leads to the check that I added out of paranoia:

```Fortran
          if (num_chars < 1) then
             ! This implies that either listDel wasn't found (in which case index will
             ! return 0) or that listDel is the next character (in which case index will
             ! return 1). Neither of these should happen: the first case implies an
             ! inconsistency with shr_string_listGetNum, and the second case should have
             ! been caught by shr_string_listIsValid. Nevertheless, we check for these
             ! possibilities here so that we can issue a meaningful error message if an
             ! issue somehow slipped through the cracks.
             call shr_string_abort(subName//" ERROR: internal inconsistency processing list = "//trim(list))
          end if
```

This leads to *another* pFUnit exception being raised, and I can't figure out a way to tell pFUnit to expect that two different exceptions were raised... I tried adding this:
```
   @assertExceptionRaised("ABORTED: (shr_string_listGetAllNames)  ERROR: internal inconsistency processing list = abc::def")
```

but that didn't get the test to pass.

There are some possible solutions here. The simplest is to add a `return` after the `shr_string_abort`; another is to change the error handling so that we don't abort at all but just return a bad rc and rely on the caller to do error checking. The first solution feels confusing, since in production this is unnecessary; and the second puts more burden on the caller. So I don't really like either of these solutions and my personal feeling is that it isn't worth complicating the production code for the sake of allowing this extra expected-error unit test: In this case, I'm comfortable reasoning that I have unit tested `shr_string_listIsValid`, so I'm not too concerned that I'm unit testing the straightforward *use* of that function followed by an abort call.

I don't feel strongly about this, though, and if you'd like to see one of these solutions - or something else - I'm open to going that direction.

**Comment by billsacks** [comment]
Thanks a lot for your review, @nusbaume ! I accepted your suggested changes – thank you for those. I have responded to your other comments, with my explanations for why I propose not changing things; I'm very open to discussing those further if you'd like. (The pFUnit expected-error handling is something that I've had extensive discussions about both with CTSM developers - especially Adrianna - and with Tom Clune (the original developer of pFUnit)... I'd be happy to talk more with you about that if you want.)

File: src/shr_string_mod.F90 (hunk 3 above)
Comment (nusbaume) [reply src/shr_string_mod.F90:1071]: Thanks for the extra context!  I am happy to skip the `stat` check if most of the other share code doesn't do it either, especially for something small and user-provided (to be honest I would be semi-impressed if a user triggered an allocation failure here).

File: test/unit/shr_string_test/test_shr_string.pf (hunk 4 above)
Comment (nusbaume) [reply test/unit/shr_string_test/test_shr_string.pf:200]: Thanks for the detailed explanation!  I didn't realize pFUnit had those limitations when it comes to testing "exceptions" (which is great to know as we are starting to use pFUnit in CAM-SIMA), so I completely agree that we should not worry about adding the extra test as none of the work-arounds seem worth just getting one additional error message tested.  I'll go ahead and mark this resolved!

File: src/shr_string_mod.F90 [hunk 5]
```diff
@@ -1011,6 +1012,106 @@ subroutine shr_string_listGetName(list,k,name,rc)
...
+
+          if (num_chars < 1) then
+             ! This implies that either listDel wasn't found (in which case index will
+             ! return 0) or that listDel is the next character (in which case index will
+             ! return 1). Neither of these should happen: the first case implies an
+             ! inconsistency with shr_string_listGetNum, and the second case should have
+             ! been caught by shr_string_listIsValid. Nevertheless, we check for these
+             ! possibilities here so that we can issue a meaningful error message if an
+             ! issue somehow slipped through the cracks.
+             call shr_string_abort(subName//" ERROR: internal inconsistency processing list = "//trim(list))
```
Comment (nusbaume) [inline src/shr_string_mod.F90]: Add list markers here as well? 




```suggestion
             call shr_string_abort(subName//" ERROR: internal inconsistency processing list = <"//trim(list)//">")
```

Apologies for missing this in my first review!

**Review by nusbaume — APPROVED** [review]
Thanks for the revisions and discussion @billsacks!  I found one last location that could use list markers, but otherwise everything looks great to me.  Thanks again!

File: src/shr_string_mod.F90 (hunk 5 above)
Comment (billsacks) [reply src/shr_string_mod.F90]: Thanks for this additional catch!
